tofu/preprocess.py

```diff
@@ -157,8 +158,25 @@ def create_phase_retrieval_pipeline(args, graph, processing_node=None):
     graph.connect_nodes(fft_phase_retrieve, phase_retrieve)
     graph.connect_nodes(phase_retrieve, ifft_phase_retrieve)
     graph.connect_nodes(ifft_phase_retrieve, crop_phase_retrieve)
-
-    return (pad_phase_retrieve, crop_phase_retrieve)
+    if args.retrieval_method == 'tie':
+        # Take the logarithm to obtain the projected thickness
+        calculate = get_task('calculate', processing_node=processing_node)
+        expression = ''
+        if args.fix_nan_and_inf:
+            expression += '(isinf (v) || isnan (v) || (v <= 0)) ? 0.0f :'
+        expression += '- log (v)'
+        if args.delta is not None:
+            import numpy as np
+            lam = 6.62606896e-34 * 299792458 / (args.energy * 1.60217733e-16)
+            # Compute mju from the fact that beta = 10^-regularization_rate * delta
+            # and mju = 4 * Pi * beta / lambda
+            mju = 4 * np.pi * 10 ** -args.regularization_rate * args.delta / lam
+            expression = '-log ({} * v) * {}'.format(2 / 10 ** args.regularization_rate, 1 / mju)
```

Comment:
The first parameter of `format` should rather be fixed in `ufo-filters`. I will wait what tomorrow's discussion brings and might shift it there.

Reply:
The code as it is now gives the projected thickness which can be directly plugged into e.g. tomographic reconstruction.

Reply:
@matze don't mind the `expression +=` vs `expression = ` inconsistency, the code is anyway going to change tomorrow. This is just to show the procedure which needs to be carried out to get the projected thickness.